src/computability/NFA.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson
+-/
+
+import data.fintype.basic
+import computability.DFA
+
+/-!
+# Nondeterministic Finite Automata
+This file contains the definition of a Nondeterministic Finite Automata (NFA), a state machine which
+determines whether a string (implemented as a list over an arbitrary alphabet) is in a regular set
+by evaluating  the string over every possible path.
+We show that DFA's are equivalent to NFA's however the construction from NFA to DFA uses an
+exponential number of states.
+-/
+
+universes u v
+
+/-- A DFA is a set of states (`state`), a transition function from state to state labelled by the
+  alphabet (`step`), a starting state (`start`) and a set of acceptance states (`accept`).
+  Note the transition function sends a state to a `finset` of states. These are the states that it
+  may be sent to. -/
+structure NFA (alphabet : Type u) :=
+(state : Type v)
+[state_fintype : fintype state]
+[state_dec : decidable_eq state]
+(step : state → alphabet → finset state)
+(start : finset state)
+(accept : finset state)
```

Comment:
It's unlikely you need both `fintype state` and `finset state`, since a set of elements taken from a finite type is naturally finite. I don't have a feeling for which is easier to remove.